fixed-hash/src/hash.rs

```diff
@@ -16,157 +20,224 @@ pub fn clean_0x(s: &str) -> &str {
-/// Takes the name of the type and the size in bytes and an optional third argument for meta data
-/// Example: `construct_hash!(H256, 32);`
-/// Example: `construct_hash!(H160, 20, cfg_attr(feature = "serialize", derive(Serialize, Deserialize)));`
+///
+/// # Examples
+///
+/// Create a public unformatted hash type with 32 bytes size.
+///
+/// ```
+/// # #[macro_use] extern crate fixed_hash;
+/// construct_hash!{ pub struct H256(32); }
+/// # fn main() {
+/// # 	assert_eq!(::std::mem::size_of::<H256>(), 32);
+/// # }
+/// ```
+///
+/// With additional attributes and doc comments.
+///
+/// ```
+/// # #[macro_use] extern crate fixed_hash;
+/// construct_hash!{
+/// 	/// My unformatted 160 bytes sized hash type.
+/// 	#[cfg_attr(feature = "serialize", derive(Serialize, Deserialize))]
+/// 	pub struct H160(20);
+/// }
+/// # fn main() {
+/// # 	assert_eq!(::std::mem::size_of::<H160>(), 20);
+/// # }
+/// ```
+///
+/// As always, visibility modifier is optional.
+/// So for a private type go ahead with this.
+///
+/// ```
+/// # #[macro_use] extern crate fixed_hash;
+/// construct_hash!{ struct H512(64); }
+/// # fn main() {
+/// # 	assert_eq!(::std::mem::size_of::<H512>(), 64);
+/// # }
+/// ```
 #[macro_export]
 macro_rules! construct_hash {
-	($from: ident, $size: expr $(, $m:meta)*) => {
+	( $(#[$attr:meta])* $visibility:vis struct $name:ident ( $n_bytes:expr ); ) => {
 		#[repr(C)]
-		$(#[$m])*
-		/// Unformatted binary data of fixed length.
-		pub struct $from (pub [u8; $size]);
+		$(#[$attr])*
+		$visibility struct $name (pub [u8; $n_bytes]);
 
-
-		impl From<[u8; $size]> for $from {
-			fn from(bytes: [u8; $size]) -> Self {
-				$from(bytes)
-			}
-		}
-
-		impl From<$from> for [u8; $size] {
-			fn from(s: $from) -> Self {
-				s.0
+		impl From<[u8; $n_bytes]> for $name {
+			#[inline]
+			fn from(bytes: [u8; $n_bytes]) -> Self {
+				$name(bytes)
 			}
 		}
 
-		impl ::core::ops::Deref for $from {
-			type Target = [u8];
-
+		impl From<$name> for [u8; $n_bytes] {
 			#[inline]
-			fn deref(&self) -> &[u8] {
-				&self.0
+			fn from(s: $name) -> Self {
+				s.0
 			}
 		}
 
-		impl AsRef<[u8]> for $from {
+		impl AsRef<[u8]> for $name {
 			#[inline]
 			fn as_ref(&self) -> &[u8] {
-				&self.0
+				self.as_bytes()
 			}
 		}
 
-		impl AsMut<[u8]> for $from {
+		impl AsMut<[u8]> for $name {
 			#[inline]
 			fn as_mut(&mut self) -> &mut [u8] {
-				&mut self.0
+				self.as_bytes_mut()
 			}
 		}
 
-		impl AsRef<$from> for $from {
+		impl $name {
+			/// Create a new, zero-initialised, instance.
+			#[deprecated(
+				since = "0.3.0",
+				note = "use the `zero` constructor for more clarity"
+			)]
 			#[inline]
-			fn as_ref(&self) -> &$from {
-				&self
+			pub fn new() -> $name {
+				Self::zero()
 			}
-		}
 
-		impl ::core::ops::DerefMut for $from {
+			/// Synonym for `new()`. Prefer to new as it's more readable.
 			#[inline]
-			fn deref_mut(&mut self) -> &mut [u8] {
-				&mut self.0
+			pub fn zero() -> $name {
```

Comment:
`$from` -> `$name` makes a lot of sense! The `$from` variable is veeery confusing, indeed..